pkg/lib/credentials.js

```diff
@@ -180,78 +180,78 @@ function Keys() {
+                    failure = _("Failed to change password");
+                    sent_new = true;
+                    return;
+                }
             }
 
-            // Exactly one of new_type or old_pass must be given
-            console.assert((new_type == null) != (old_pass == null));
-
-            const cmd = ["ssh-keygen", "-f", file];
-            if (new_type)
-                cmd.push("-t", new_type);
-            else
-                cmd.push("-p");
-
-            const proc = cockpit.spawn(cmd, { pty: true, environ: ["LC_ALL=C"], err: "out", directory: self.path });
-
-            const timeout = window.setTimeout(() => {
-                failure = _("Prompting via ssh-keygen timed out");
-                proc.close("terminated");
-            }, 10 * 1000);
-
-            proc
-                    .stream(data => {
-                        buffer += data;
-                        if (old_pass) {
-                            for (let i = 0; i < old_exps.length; i++) {
-                                if (old_exps[i].test(buffer)) {
-                                    buffer = "";
-                                    failure = _("Old password not accepted");
-                                    proc.input(old_pass + "\n", true);
-                                    return;
-                                }
-                            }
-                        }
-
-                        for (let i = 0; i < new_exps.length; i++) {
-                            if (new_exps[i].test(buffer)) {
-                                buffer = "";
-                                proc.input(new_pass + "\n", true);
-                                failure = _("Failed to change password");
-                                sent_new = true;
-                                return;
-                            }
-                        }
-
-                        if (sent_new) {
-                            for (let i = 0; i < bad_exps.length; i++) {
-                                if (bad_exps[i].test(buffer)) {
-                                    failure = _("New password was not accepted");
-                                    return;
-                                }
-                            }
-                        }
-                    })
-                    .then(resolve)
-                    .catch(ex => {
-                        if (ex.exit_status)
-                            ex = new Error(failure);
-                        reject(ex);
-                    })
-                    .finally(() => window.clearInterval(timeout));
+            if (sent_new) {
+                for (let i = 0; i < bad_exps.length; i++) {
+                    if (bad_exps[i].test(buffer)) {
+                        failure = _("New password was not accepted");
+                        return;
+                    }
+                }
+            }
         });
+
+        try {
+            await proc;
```

Comment:
The `.stream()` function returns immediately, right?  And we effectively end up waiting on `await proc` directly.  Why not move the `timeout` code down here too, so that it's closer to the thing we're actually waiting on, and closer to the cancellation inside of the `finally` block?

Reply:
> The .stream() function returns immediately, right? And we effectively end up waiting on await proc directly.

Correct. `.stream()` chains, it just returns the original promise (`return this`). I just find it easier to read like that. It avoids putting the whole `.stream` block into the `try {}` -- exceptions inside of that are really not handled by that try/catch.

> Why not move the timeout code down here too

Can do.

Reply:
Moved.